File: road-system/typescript/src/roadSystem.js
function solution(roadRegister) {
    var numCities = roadRegister.length;
    for (var city = 0; city < numCities; city++) {
        var isRoadSystemConvenient = getIncomingRoads(city, roadRegister) === getOutgoingRoads(city, roadRegister);
        if (!isRoadSystemConvenient) {
            return false;
        }
    }
    return true;
}
function getIncomingRoads(city, roadRegister) {
    var numIncoming = 0;
    for (var row = 0; row < roadRegister.length; row++) {
        if (roadRegister[row][city]) {
            numIncoming++;
        }
    }
    return numIncoming;
}
function getOutgoingRoads(city, roadRegister) {
    var numOutgoing = 0;
    for (var col = 0; col < roadRegister.length; col++) {
        if (roadRegister[city][col]) {
            numOutgoing++;
        }
    }
    return numOutgoing;
}

File: road-system/ruby/src/RoadSystem.rb
def solution(road_register)
  num_cities = road_register.length

  (0...num_cities).each do |city|
    return false if get_incoming_roads(city, road_register) != get_outgoing_roads(city, road_register)
  end

  true
end

def get_incoming_roads(city, road_register)
  road_register.transpose[city].count(true)
end

def get_outgoing_roads(city, road_register)
  road_register[city].count(true)
end
